taichi/backends/metal/shaders/print.metal.h

```diff
@@ -0,0 +1,146 @@
+        int num_entries) {
+      return (num_entries + kMetalNumPrintMsgTypePerI32 - 1) /
+             kMetalNumPrintMsgTypePerI32;
+    }
+
+    [[maybe_unused]] inline int mtl_compute_print_msg_bytes(int num_entries) {
+      // First int32: stores |num_entries|
+      // Then follows the number of int32's to store the type masks.
+      // Finally the print data. 4 byte for each entry.
+      const int sz =
+          sizeof(int32_t) *
+          (1 + mtl_compute_num_print_msg_typemasks(num_entries) + num_entries);
+      return sz;
+    }
+
+    class PrintMsg {
+     public:
+      // Data layout:
+      //
+      // * 1 i32 to record how many args there are
+      // * Followed by M i32s, which are the type masks. Each i32 can encode
+      // hold up to |kMetalNumPrintMsgTypePerI32| types.
+      // * Followed by N i32s, one for each print arg. F32 are encoded to I32.
+      // For strings, there is a string table on the host side, so that the
+      // kernel only needs to store a I32 string ID.
+      enum Type { I32 = 1, F32 = 2, Str = 3 };
+
+      PrintMsg(device int32_t * buf, int num_entries)
+          : mask_buf_(buf),
+            data_buf_(buf + mtl_compute_num_print_msg_typemasks(num_entries)) {
+      }
+
+      void pm_set_i32(int i, int x) {
+        set_entry(i, x, Type::I32);
+      }
+
+      void pm_set_f32(int i, float x) {
+        const int32_t ix = *reinterpret_cast<thread int32_t *>(&x);
+        set_entry(i, ix, Type::F32);
+      }
+
+      void pm_set_str(int i, int str_id) {
+        set_entry(i, str_id, Type::Str);
+      }
+
+      Type pm_get_type(int i) {
+        const int mask_i = i / kMetalNumPrintMsgTypePerI32;
+        const int i_in_mask = i % kMetalNumPrintMsgTypePerI32;
+        int mask = mask_buf_[mask_i];
+        mask >>= typemask_shift(i_in_mask);
+        mask &= kMetalPrintMsgTypeWidthMask;
+        return (Type)mask;
+      }
+
+      int32_t pm_get_data(int i) {
+        return data_buf_[i];
+      }
+
+     private:
+      void set_entry(int i, int32_t x, Type ty) {
+        const int mask_i = i / kMetalNumPrintMsgTypePerI32;
+        const int i_in_mask = i % kMetalNumPrintMsgTypePerI32;
+        int mask = ((int)ty & kMetalPrintMsgTypeWidthMask);
+        mask <<= typemask_shift(i_in_mask);
+        mask_buf_[mask_i] |= mask;
+        data_buf_[i] = x;
+      }
+
+      inline static int typemask_shift(int i_in_mask) {
+        return (kMetalNumPrintMsgTypePerI32 - 1 - i_in_mask) *
+               kMetalNumBitsPerPrintMsgType;
+      }
+
+      device int32_t *mask_buf_;
+      device int32_t *data_buf_;
+    };
+
+    struct PrintMsgAllocator { atomic_int next; };
+
+    constant constexpr int kMetalPrintBufferSize =
+        2 * 1024 * 1024 - sizeof(PrintMsgAllocator);  // 2MB
+
+    [[maybe_unused]] device int32_t *
+    mtl_print_alloc_buf(device PrintMsgAllocator * pa, int num_entries) {
+      const int sz = mtl_compute_print_msg_bytes(num_entries);
```

Comment:
IIUC you are using dynamic size instead of fixed size for each `print` like OpenGL does? Cool!

Reply:
Yep, Metal's syntax really helped simplify the design here..